2018/proposals/gensim Aneesh Joshi Similarity Learning.md

```diff
@@ -0,0 +1,399 @@
+  </tr>
+  <tr>
+  	 <td align="center", bgcolor=#eeeeee> K-NRM </td>
+  	 <td align="center", bgcolor=#eeeeee> 0.6268 </td>
+  	 <td align="center", bgcolor=#eeeeee> 0.6693 </td>
+  	 <td align="center", bgcolor=#eeeeee> 0.6256 </td>
+  </tr>
+</table>
+
+Along with the loss:
+
+![alt-text](https://github.com/faneshion/MatchZoo/blob/master/docs/_static/images/matchzoo.wikiqa.loss.png)
+
+Free datasets for Similarity Learning will also be added to [gensim-data](https://github.com/RaRe-Technologies/gensim-data) :
+
+* [Quora Duplicate Questions](https://data.quora.com/First-Quora-Dataset-Release-Question-Pairs) (already in gensim-data)
+* [WikiQA](https://www.microsoft.com/en-us/download/details.aspx?id=52419)
+* [Ubuntu Dialogue Corpus](https://github.com/rkadlec/ubuntu-ranking-dataset-creator)
+
+# Deliverables
+
+The whole project can be broken down into 4 milestones
+1. Benchmark and Evaluation of existing models
+2. Get one model working end-to-end
+3. Repeat the same for remaining models
+4. Disseminate models
+
+### Community Bonding Period (April 23 to May 14)
+* Design the api for the models in gensim (Basically what goes where and how the user calls it)
+* Discuss and update the list of models to be implemented.
+* Start verifying and benchmarking all the mentioned models using [MatchZoo](https://github.com/faneshion/MatchZoo)
+* Publish results and comparisons as a blog
+
+**Milestone 1 complete!**
+
+### Week 1 (May 14 to May 20)
+* Find and understand existing data preprocessing code in gensim
+* Develop the code for similarity learning specific data preprocessing
+* Add data streaming wherever possible
+
+### Week 2 (May 21 to May 27)
+* Stitch existing and new code together into a module for Similarity Learning related Preprocessing
+* Write Test cases for Data Preprocessing module
+* Write Documentation for the module-
+
+I will first make a concrete and full implementation of DSSM and then apply the same process to the remaining models
+
+### Week 3 (May 28 to June 3)
+* Evaluate and take feedback on existing DSSM code for possible improvements
+* Make improvements to code(if required)
+* Add dependencies for installation and continuous integration
+* Add model serializability and other such gensim supports
+
+### Week 4 (June 4 to June 10)
+* Add test cases
+* Add documentation
+* Write a blog describing the new model
+* Merge my code into `gensim/develop`
+
+**Milestone 2 complete!** <br/>
+At this point, a full end-to-end DSSM model will be ready which will set the precedent for further models.
+
+### Midterm Evaluation 1 (June 11 to June 15)
+* Work on feedback
+* Catch up on any unseen work
+
+The final list on models might have changed by the time I reach this point.
+This week onwards, I will work on 2/3 models per week. If the number of models is more, I will increase the rate.
+Assuming the list remains the same, here is what I'll do:
+
+### Week 5 (June 16 to June 24)

```

Comment:
so, need to re-work timeline :( 

General scheme:
0. Setup evaluation pipeline
1. Collect/implement models for comparison, tune parameters, benchmark it
2. Choose the best model, integrate it to gensim (make similar API, add streaming support, tests, documentation)
3. Work on improving model (collect tips for tuning, maybe some additional thing that improves a quality, etc).